File: production/pre_processing/TextProcessor.py
#!/usr/bin/env python
# coding: utf-8

# In[2]:


from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.tokenize import RegexpTokenizer
from nltk.stem import SnowballStemmer

from typing import List
import time
import re

import sys
sys.path.append('../')  # Go up two folders to the project root
from utilities.General_Utilities import read_words_from_file 


# In[6]:


"""
    This class is used in the project to implement pre-processing text elaboration.
    It contains several methods for text cleaning, stemming and stop word removal.
"""
class TextProcessor:
    
    def __init__(self, use_stemming_and_stop_words_removal:bool):
        self.use_stemming_and_stop_words = use_stemming_and_stop_words_removal # flag to enable stemming and stop_words_removal
        
        #Regular expression used to clean a raw text.
        self.reg_exp_punctuation = r'[^\w\s]|_'
        self.reg_exp_html_tags=r'<[^>]+>'
        self.reg_exp_hashtags = r'#\w+' 
        self.reg_exp_usernames = r'@\w+'
        self.control_char_pattern=r'[^\x00-\x7F]+'
        self.reg_exp_web_link_pattern=r'https*://\S+|www.\S+'
        
        #A faster an improved version of PorterStemmer
        self.stemmer = SnowballStemmer("english")
        
        #Initialize the list of all English stopwords
        #Use the default NLTK stopword + additional stopwords find on the web and grouped in a single file.
        stop_word_list=[]
        try:
            stop_word_list=read_words_from_file("../utilities/english_stop_words.txt")
        except Exception as e:   
            print("Stopword file english_stop_words.txt not imported")
        self.stop_words=set(stopwords.words("english")+stop_word_list)
        
        #print(self.stop_words)
        #print(len(self.stop_words))
            
    def process_text(self, text:str, return_tokens = False)->str:
        """ Process a text by cleaning, tokenizing, removing stopwords and stemming(optionally).
            This is the main method called from outside, during the processing of all the documents.

            Args:
                text: The text to be processed.
            Returns:
                The processed text after cleaning, tokenizing, removing stopwords, and stemming.
        """
        
        # Remove special characters
        text = self.clean_text(text)
        
        # To handle case of empty string
        if not text:
            return text

        # Transform the document in a list of tokens
        tokenizer = RegexpTokenizer(r'\w+')
        word_tokens = tokenizer.tokenize(text)

        # Remove stopwords and apply stemming.
        if self.use_stemming_and_stop_words:
            word_tokens = self.remove_stopwords(word_tokens)
            word_tokens = self.stem_text(word_tokens)

        # Transform and return the list of tokens in a document (unique string)
        if return_tokens == True:
            return word_tokens
            
        return ' '.join(word_tokens)

    def stem_text(self, tokens:List[str])->List[str]:
        """Apply Stemming on a list of tokens using English language.
        
        Args:
            tokens: List of tokens to be stemmed.
        Returns:
            The list of tokens after stemming based on the provided language. 
        """
        
        # Do stem for each tokens in the list
        return [self.stemmer.stem(word) for word in tokens]

    def remove_stopwords(self, tokens:List[str]):
        """Remove stopwords from a list of stop words tokens in English language.

        Args:
            tokens: List of tokens to be processed.
            language: The language used to identify the stopwords.
        Returns:
            The list of tokens after removing the stopwords based on the provided language. 
            If the language is not supported, the function returns the original tokens.
        """

        # If the tokens is in the list of stopwords, remove it
        return [word for word in tokens if word not in self.stop_words]
    
    def clean_text(self,text:str)->str:
        """Clean the text by converting to lowercase, replacing special characters.

        Args:
            text: The text to be cleaned.
        Returns:
            The cleaned text after converting to lowercase, replacing special characters, and removing emojis.
        """
        text = str(text).lower()

        # Replace special characters 
        text = re.sub(self.control_char_pattern, " ", text)
        
        combined_pattern = re.compile(self.reg_exp_html_tags+'|'+self.reg_exp_hashtags + '|' + self.reg_exp_punctuation + '|' + self.reg_exp_usernames + '|' + self.reg_exp_web_link_pattern)
        text = re.sub(combined_pattern, " ", text)

        #Collapse white spaces between words.
        text = re.sub(r'\s+', ' ', text)

        return text.strip()
    
    '''
    def lemmatize(self, tokens, language):
        lemmatizer = WordNetLemmatizer()
        
        lemmatized = [lemmatizer.lemmatize(w, language) 
                      for w in tokens]
        
        return lemmatized
    '''

File: production/query_processing/SearchEngine/src/main/java/it/unipi/mircv/SearchEngine/utilities/TextProcessor.java
package it.unipi.mircv.SearchEngine.utilities;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.regex.Pattern;
import java.util.stream.Collectors;


import org.tartarus.snowball.ext.englishStemmer;
import org.tartarus.snowball.SnowballStemmer;



/**
 * This class is used in the project to implement pre-processing text elaboration.
 * It contains several methods for text cleaning, stemming and stop word removal.
 * It has the same exact implementation of TextProcessor used in Python during Indexing.
 * 
 * @author Davide and Gabriele
 *
 */
public class TextProcessor {
	
    private boolean useStemmingAndStopWords;
    private final Pattern regExpPunctuation = Pattern.compile("[^\\w\\s]|_");
    private final Pattern regExpHtmlTags = Pattern.compile("<[^>]+>");
    private final Pattern regExpHashtags = Pattern.compile("#\\w+");
    private final Pattern regExpUsernames = Pattern.compile("@\\w+");
    private final Pattern controlCharPattern = Pattern.compile("[^\\x00-\\x7F]+");
    private final Pattern regExpWebLinkPattern = Pattern.compile("https*://\\S+|www.\\S+");

    private SnowballStemmer stemmer;
    private Set<String> stopWords;

    
    /**
     * Constructor method:
     * 
     * @param useStemmingAndStopWordsRemoval -  true if stemming and stop word removal should be done, false otherwise
     * @param pathOfStopWords - the file path where to load the stop words
     */
    public TextProcessor(boolean useStemmingAndStopWordsRemoval,String pathOfStopWords) {
        
    	this.useStemmingAndStopWords = useStemmingAndStopWordsRemoval;
        this.stopWords = readWordsFromFile(pathOfStopWords);
        
        if (useStemmingAndStopWordsRemoval) {
        	 englishStemmer en= new englishStemmer();
             this.stemmer = (SnowballStemmer) en;
        }
    }
    
    /**
     * Process a text by cleaning, tokenizing, removing stopwords and stemming(optionally).
     * This is the main method called from outside, during the processing of all the queries.
     * 
     * @param text - The text to be processed.
     * @return
     */
    public String processText(String text) {
        
    	text = cleanText(text);

        if (text.isEmpty()) {
            return text;
        }

        List<String> wordTokens = tokenizeText(text);

        if (useStemmingAndStopWords) {
            wordTokens = removeStopWords(wordTokens);
            wordTokens = stemText(wordTokens);
        }

        return String.join(" ", wordTokens);
    }
    
    /**
     * Apply Stemming on a list of tokens using English language.
     * @param tokens - List of tokens to be stemmed.
     * @return - a list of strings where stemming is done.
     */
    public List<String> stemText(List<String> tokens) {
//        List<String> stemmedTokens = new ArrayList<>();
//        for (String token : tokens) {
//        	
//        	stemmer.setCurrent(token);
//			stemmer.stem();
//            stemmedTokens.add(stemmer.getCurrent());
//        }
//        return stemmedTokens;
    	// JAVA8
    	return tokens.stream()
                .map(token -> {
                    stemmer.setCurrent(token);
                    stemmer.stem();
                    return stemmer.getCurrent();
                })
                .collect(Collectors.toList());
    	
    	
    }

    /**
     * Remove stopwords from a list of stop words tokens in English language.
     * @param tokens - List of tokens to be processed.
     * @return
     */
    public List<String> removeStopWords(List<String> tokens) {
        List<String> filteredTokens = new ArrayList<>();
        for (String token : tokens) {
            if (!stopWords.contains(token)) {
                filteredTokens.add(token);
            }
        }
        return filteredTokens;
        
    	//Java 8
//        return tokens.stream()
//                .filter(token -> !stopWords.contains(token))
//                .collect(Collectors.toList());
        
    }
	/**
	 * Clean the text by converting to lowercase, replacing special characters.
	 * @param text - The text to be cleaned.
	 * @return The cleaned text
	 */
    public String cleanText(String text) {
        text = text.toLowerCase();
        text = controlCharPattern.matcher(text).replaceAll(" ");

        String combinedPattern = regExpHtmlTags.pattern() + "|" + regExpHashtags.pattern() + "|" +
                regExpPunctuation.pattern() + "|" + regExpUsernames.pattern() + "|" +
                regExpWebLinkPattern.pattern();
        Pattern combinedPatternCompiled = Pattern.compile(combinedPattern);
        text = combinedPatternCompiled.matcher(text).replaceAll(" ");

        text = text.replaceAll("\\s+", " ");

        return text.trim();
    }
    
    
    /**
     * Method to divide text in tokens.
     * @param text - the text to be divided
     * @return a list of strings representing the tokens
     */
    public List<String> tokenizeText(String text) {
        String[] words = text.split("\\s+");
        return Arrays.asList(words);
    }
    
	/**
	 * Utility method for loading a list of stop words from a textual file.
	 * @param filePath
	 * @return
	 */
    private Set<String> readWordsFromFile(String filePath) {
        Set<String> stopWordsList = new HashSet<>();
        try {
            Path path = Paths.get(filePath);
            stopWordsList.addAll(Files.readAllLines(path));
        } catch (IOException e) {
            System.err.println("Stopword file not imported");
        }
        return stopWordsList;
    }
}
